# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/WebrtcCallWrapper.cpp
# Commit: 786ea95e928b
# Full Hash: 786ea95e928b7bcec53a56b50f1d0ff136c76a62
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 1
# Description:
#   Bug 1654112 - Move NoTrialsConfig into SharedWebrtcState and make WebrtcCallWrapper keep a ref to it. r=ng
#   
#   SharedWebrtcState when managed by PeerConnectionCtx is destroyed on main thread.
#   With interactions with call instances being async, there is no guarantee that
#   the objects owned by SharedWebrtcState (and referenced through raw pointers by
# ==============================================================================

diff -r 2a373d864092 -r 786ea95e928b dom/media/webrtc/libwebrtcglue/WebrtcCallWrapper.cpp
--- a/dom/media/webrtc/libwebrtcglue/WebrtcCallWrapper.cpp	Wed Apr 14 17:40:24 2021 +0200
+++ b/dom/media/webrtc/libwebrtcglue/WebrtcCallWrapper.cpp	Fri Apr 16 14:44:09 2021 +0200
@@ -15,42 +15,25 @@
 /* static */ RefPtr<WebrtcCallWrapper> WebrtcCallWrapper::Create(
     const dom::RTCStatsTimestampMaker& aTimestampMaker,
     UniquePtr<media::ShutdownBlockingTicket> aShutdownTicket,
-    const RefPtr<SharedWebrtcState>& aSharedState,
-    webrtc::WebRtcKeyValueConfig* aTrials) {
-  return Create(
-      aTimestampMaker, std::move(aShutdownTicket),
-      aSharedState->mCallWorkerThread.get(),
-      [aSharedState] { return aSharedState->GetModuleThread(); },
-      aSharedState->mAudioStateConfig, aSharedState->mAudioDecoderFactory,
-      aTrials);
-}
-
-/* static */ RefPtr<WebrtcCallWrapper> WebrtcCallWrapper::Create(
-    const dom::RTCStatsTimestampMaker& aTimestampMaker,
-    UniquePtr<media::ShutdownBlockingTicket> aShutdownTicket,
-    RefPtr<AbstractThread> aCallThread,
-    std::function<webrtc::SharedModuleThread*()> aModuleThreadGetter,
-    const webrtc::AudioState::Config& aAudioStateConfig,
-    webrtc::AudioDecoderFactory* aAudioDecoderFactory,
-    webrtc::WebRtcKeyValueConfig* aTrials) {
+    const RefPtr<SharedWebrtcState>& aSharedState) {
   auto eventLog = MakeUnique<webrtc::RtcEventLogNull>();
-  webrtc::Call::Config config(eventLog.get());
-  config.audio_state = webrtc::AudioState::Create(aAudioStateConfig);
   auto taskQueueFactory = MakeUnique<SharedThreadPoolWebRtcTaskQueueFactory>();
-  config.task_queue_factory = taskQueueFactory.get();
-  config.trials = aTrials;
   auto videoBitrateAllocatorFactory =
       WrapUnique(webrtc::CreateBuiltinVideoBitrateAllocatorFactory().release());
   RefPtr<WebrtcCallWrapper> wrapper = new WebrtcCallWrapper(
-      std::move(aCallThread), aAudioDecoderFactory,
-      std::move(videoBitrateAllocatorFactory), std::move(eventLog),
-      std::move(taskQueueFactory), aTimestampMaker, std::move(aShutdownTicket));
+      aSharedState, std::move(videoBitrateAllocatorFactory),
+      std::move(eventLog), std::move(taskQueueFactory), aTimestampMaker,
+      std::move(aShutdownTicket));
 
-  wrapper->mCallThread->Dispatch(NS_NewRunnableFunction(
-      __func__, [wrapper, config = std::move(config),
-                 moduleThreadGetter = std::move(aModuleThreadGetter)] {
-        wrapper->SetCall(
-            WrapUnique(webrtc::Call::Create(config, moduleThreadGetter())));
+  wrapper->mCallThread->Dispatch(
+      NS_NewRunnableFunction(__func__, [wrapper, aSharedState] {
+        webrtc::Call::Config config(wrapper->mEventLog.get());
+        config.audio_state =
+            webrtc::AudioState::Create(aSharedState->mAudioStateConfig);
+        config.task_queue_factory = wrapper->mTaskQueueFactory.get();
+        config.trials = aSharedState->mTrials.get();
+        wrapper->SetCall(WrapUnique(
+            webrtc::Call::Create(config, aSharedState->GetModuleThread())));
       }));
 
   return wrapper;
@@ -106,18 +89,18 @@
 WebrtcCallWrapper::~WebrtcCallWrapper() { MOZ_ASSERT(!mCall); }
 
 WebrtcCallWrapper::WebrtcCallWrapper(
-    RefPtr<AbstractThread> aCallThread,
-    RefPtr<webrtc::AudioDecoderFactory> aAudioDecoderFactory,
+    RefPtr<SharedWebrtcState> aSharedState,
     UniquePtr<webrtc::VideoBitrateAllocatorFactory>
         aVideoBitrateAllocatorFactory,
     UniquePtr<webrtc::RtcEventLog> aEventLog,
     UniquePtr<webrtc::TaskQueueFactory> aTaskQueueFactory,
     const dom::RTCStatsTimestampMaker& aTimestampMaker,
     UniquePtr<media::ShutdownBlockingTicket> aShutdownTicket)
-    : mTimestampMaker(aTimestampMaker),
+    : mSharedState(std::move(aSharedState)),
+      mTimestampMaker(aTimestampMaker),
       mShutdownTicket(std::move(aShutdownTicket)),
-      mCallThread(std::move(aCallThread)),
-      mAudioDecoderFactory(std::move(aAudioDecoderFactory)),
+      mCallThread(mSharedState->mCallWorkerThread),
+      mAudioDecoderFactory(mSharedState->mAudioDecoderFactory),
       mVideoBitrateAllocatorFactory(std::move(aVideoBitrateAllocatorFactory)),
       mEventLog(std::move(aEventLog)),
       mTaskQueueFactory(std::move(aTaskQueueFactory)) {}